File: extras/aethon/android/app/src/main/java/hev/htproxy/TProxyService.java
package hev.htproxy;

/** Exact JNI contract exported by the pinned HEV Socks5 Tunnel 2.16.0 source. */
public final class TProxyService {
    static {
        System.loadLibrary("hev-socks5-tunnel");
    }

    private TProxyService() { }

    public static native void TProxyStartService(String configPath, int fd);
    public static native void TProxyStopService();
    public static native long[] TProxyGetStats();
}

File: android/app/src/main/java/hev/htproxy/TProxyService.kt
package hev.htproxy

/**
 * JNI facade expected by `libhev-socks5-tunnel.so` ([JNI_OnLoad] registers these natives
 * against `hev/htproxy/TProxyService`).
 *
 * Library load is deferred and failure-soft so a missing/misaligned `.so` cannot crash
 * process startup on 16 KB page-size devices.
 */
object TProxyService {
    @Volatile private var loaded = false
    @Volatile private var loadError: String? = null

    fun ensureLoaded(): Boolean {
        if (loaded) return true
        synchronized(this) {
            if (loaded) return true
            return try {
                System.loadLibrary("hev-socks5-tunnel")
                loaded = true
                true
            } catch (t: Throwable) {
                loadError = t.message ?: t.javaClass.simpleName
                false
            }
        }
    }

    fun loadError(): String? = loadError

    external fun TProxyStartService(configPath: String, fd: Int)
    external fun TProxyStopService()
    external fun TProxyGetStats(): LongArray
}
